Mathport/Syntax/Translate/Tactic/Mathlib/Misc1.lean

```diff
@@ -189,44 +189,46 @@ open AST3 Parser
+  let o := optTk o
+  let hs := (← trSimpArgs args).asNonempty
+  let colon := optTk !attrs.isEmpty
+  let attrs := attrs.map mkIdent |>.asNonempty
+  `(command| #simp $[only%$o]? $[[$hs,*]]? $[with $attrs*]? $[:%$colon]? $(← trExpr e))
 
 -- # tactic.simp_result
 @[trTactic dsimp_result] def trDSimpResult : TacM Syntax := do
-  let o := if ← parse onlyFlag then mkNullNode #[mkAtom "only"] else mkNullNode
-  let hs := trSimpList (← trSimpArgs (← parse simpArgList))
-  pure $ mkNode ``Parser.Tactic.dsimpResult #[mkAtom "dsimp_result", o, hs,
-    trSimpAttrs $ (← parse (tk "with" *> ident*)?).getD #[],
-    mkAtom "=>", ← trBlock (← itactic)]
+  let o := optTk (← parse onlyFlag)
+  let hs := (← trSimpArgs (← parse simpArgList)).asNonempty
+  let attrs := (← parse (tk "with" *> ident*)?).getD #[] |>.map mkIdent |>.asNonempty
+  `(tactic| dsimp_result $[only%$o]? $[[$hs,*]]? $[with $attrs*]? => $(← trBlock (← itactic)))
 
 @[trTactic simp_result] def trSimpResult : TacM Syntax := do
-  let o := if ← parse onlyFlag then mkNullNode #[mkAtom "only"] else mkNullNode
-  let hs := trSimpList (← trSimpArgs (← parse simpArgList))
-  pure $ mkNode ``Parser.Tactic.simpResult #[mkAtom "simp_result", o, hs,
-    trSimpAttrs $ (← parse (tk "with" *> ident*)?).getD #[],
-    mkAtom "=>", ← trBlock (← itactic)]
+  let o := optTk (← parse onlyFlag)
+  let hs := (← trSimpArgs (← parse simpArgList)).asNonempty
+  let attrs := (← parse (tk "with" *> ident*)?).getD #[] |>.map mkIdent |>.asNonempty
+  `(tactic| simp_result $[only%$o]? $[[$hs,*]]? $[with $attrs*]? => $(← trBlock (← itactic)))
 
 -- # tactic.simpa
 @[trTactic simpa] def trSimpa : TacM Syntax := do
-  let unfold := (← parse (tk "!")?).map fun _ => default
-  let squeeze := (← parse (tk "?")?).map fun _ => default
-  let o := if ← parse onlyFlag then some default else none
-  let hs := some (← trSimpArgs (← parse simpArgList)) |>.filter (!·.isEmpty)
-  let attrs := (← parse (tk "with" *> ident*)?).map (·.map mkIdent)
+  let unfold ← parse (tk "!")?; let squeeze ← parse (tk "?")?
+  let o := optTk (← parse onlyFlag)
+  let hs := (← trSimpArgs (← parse simpArgList)).asNonempty
+  let attrs := (← parse (tk "with" *> ident*)?).getD #[] |>.map mkIdent |>.asNonempty
   let e ← liftM $ (← parse (tk "using" *> pExpr)?).mapM trExpr
-  let (cfg, disch) ← parseSimpConfigCore (← expr?)
-  let cfg := cfg.bind quoteSimpConfig
-  `(tactic| simpa $[!%$unfold]? $[?%$squeeze]? $[(config := $cfg)]? $[(disch := $disch:tactic)]?
-      $[only%$o]? $[[$hs,*]]? $[with $attrs*]? $[using $e]?)
+  let (cfg, disch) ← parseSimpConfig (← expr?)
+  let cfg ← mkConfigStx? (cfg.bind quoteSimpConfig)
+  let rest ← `(Mathlib.Tactic.simpaArgsRest|
+    $[$cfg:config]? $(disch)? $[only%$o]? $[[$hs,*]]? $[with $attrs*]? $[using $e]?)
+  match unfold, squeeze with
+  | none, none => `(tactic| simpa $rest)
+  | none, some _ => `(tactic| simpa? $rest)
+  | some _, none => `(tactic| simpa! $rest)
+  | some _, some _ => `(tactic| simpa!? $rest)
```

Comment:
@gebner You will be sad to see that I partially reverted your change here. The reason is because the formatter is too smart for its own good and sees that `"simpa" "!" "?"` should be rendered as `simpa !?` even though there is no space specified, so we have to explicitly select these alternate tactics if we want the short rendering (and I do).

Reply:
I can't say that I'm excited about the change, but at least we're on the same page about the formatting: every space counts.